File: source/20221120_20221127/Ruby/single_file/q_01_shortest_substring_with_small_and_capital_forms_of_letters.rb
#!/usr/bin/ruby

=begin
This script with the function 'solution' given
a string input computes the length of the shortest 
substring that only accommodates both small and capital
versions of all the letters included in the indicated
substring.
=end

=begin
  Worked on this code on 
                        Dec 12, 2022 / Mon
=end

require 'set'

def solution(s_input)
  sz = s_input.length
=begin
  Obtained size.
=end
      
  smalls   = ("a".."z").to_a
  capitals = ("A".."Z").to_a
=begin
  Got the arrays of smalls and capitals.
=end
  
#   puts smalls.join(" ")
#   puts capitals.join(" ")
  
  mp_small = Hash.new
  mp_cap   = Hash.new
  
  for kk in 0..(-1+smalls.length) do
    mp_small[smalls[kk]]   = capitals[kk]
    mp_cap  [capitals[kk]] = smalls[kk]
  end
=begin
  Created hashes for pointers between
  corresponding characters.
=end
  
#   for key , value in mp_small
#     puts "#{key} -> #{value}"
#   end
  
  for s_length in 2..sz
    for s_index_start in 0..(sz-s_length)
=begin
        A useful substring has a length starting from 2
        and extending to the length of the whole string.
        
        The starting index for the substring in question
        has to have a range as indicated by the statement
        defining the iteration index for the inner 'for'
        loop.
        
        Since 's_length' is iteratively set starting
        from the lowest length possible, the first
        success to be encountered during the search
        will mark the time to draw the function to
        an end, having found the shortest string that
        accommodates nothing other than some characters
        in both small and capital forms.
=end
      temp = s_input[s_index_start..(s_index_start+s_length-1)]
      temp_set = Set.new(temp.chars)
=begin
      Created a set out of the characters in 'temp'.
      No duplicates allowed.
=end
      
#       puts temp
#       puts (temp_set.to_a).join(" ")
      
      flag_found = true
=begin
        If the flag above remains true after the 'for' loop
        below, then it will be time to return and yield
        the result.
=end
      
      temp_set.each do |item|
        if ( mp_small.key?(item) and temp_set.include?( mp_small[item] ) ) or \
           ( mp_cap.key?(item)   and temp_set.include?( mp_cap[item] ) )
        else
          flag_found = false
          break
=begin
            If the condition above is not satisfied even for
            a single character, then we should break out of the
            for loop since there is no need to lose time
            checking the remaining characters. Setting the
            flag to false will prevent the remaining 'if'
            conditional from terminating the function.
=end
        end # if
      end # each item
=begin
      Checks for element in hash or set in each iteration
      of the loop above.
=end
      
      if flag_found
        puts "String found in \'#{s_input}\' :\n  #{temp}"
        return s_length
      end
=begin
        The first success to be encountered has the function
        return the length of the string in question.
=end
    end # for s_index_start
  end # for s_length
  
  return -1
=begin
    No possible success meaning there is no substring within
    the original string accommodating both small and capital
    forms of some letters, and nothing else.
=end
end


list_of_strings = [
  "Bb" ,
  "aACCCxA" ,
  "CCzzcZazAxAZ" ,
  "CCzzcybzAxAZ" ]

list_of_strings.each do |s_input|
  puts ""
  
  response = solution(s_input)
  puts "Response to \"#{s_input}\" :"
  puts "  " + String(response)
end

File: source/20221120_20221127/Lua/single_file/q_01_shortest_substring_with_small_and_capital_forms_of_letters.lua
#!/usr/bin/lua

--[=====[
This script with the function 'solution' given
a string input computes the length of the shortest 
substring that only accommodates both small and capital
versions of all the letters included in the indicated
substring.
--]=====]

--[=====[
  Worked on this code on 
                        Dec 16, 2022 / Fri
--]=====]

--[=====[
  Function to compute a range of characters in the
  ASCII table. I obtained this function from a webpage.
--]=====]
function range_char(c_start, c_stop)
  local s_list = {}
  
  for kk = 0 , ( c_stop:byte() - c_start:byte() ) do
    table.insert( s_list, string.char( c_start:byte() + kk ) )
  end -- for kk
  
  return s_list
end

--[=====[
  See the answer at:
    https://stackoverflow.com/questions/20066835/lua-remove-duplicate-elements
--]=====]
function remove_duplicates(array_input)
  local hash = {}
--   local res = {}

  for _ , v in ipairs(array_input) do
    if (not hash[v]) then
--       res[#res+1] = v
      hash[v] = true
    end
  end
  
  return hash
end -- function remove_duplicates

function solution(A)
  local smalls   = range_char( 'a' , 'z' )
  local capitals = range_char( 'A' , 'Z' )
  --[=====[
    Make use of the function 'range_char' to
    compute separately the ranges for the small
    and capital letters. These are now available
    as tables.
  --]=====]
  
  local mp_small = {}
  local mp_cap   = {}
  
  for kk = 1 , #smalls do
    mp_small[smalls[kk]]   = capitals[kk]
    mp_cap  [capitals[kk]] = smalls[kk]
  end -- for kk
  --[=====[
    Set up hashes to map each of the small
    letters to the corresponding capital letter,
    and vice versa. Hashes will save us time through
    the virtually constant-time access facility.
  --]=====]
  
  --[=====[
  for key , val in pairs(mp_small) do
    print( string.format( "  %s -> %s" , key , mp_small[key] ) )
  end -- for key val
  --]=====]
  
  local s_list = {}
  for s in A:gmatch(".") do
    table.insert(s_list, s)
  end
  local sz = string.len(A)
  --[=====[
    Obtain the table of characters within the given
    string A.
  --]=====]
  
  for s_length = 2 , sz do
    for s_index_start = 1 , (sz - s_length + 1) do
      --[=====[
        A useful substring has a length starting from 2
        and extending to the length of the whole string.
        
        The starting index for the substring in question
        has to have a range as indicated by the statement
        defining the iteration index for the inner 'for'
        loop.
        
        Since 's_length' is iteratively set starting
        from the lowest length possible, the first
        success to be encountered during the search
        will mark the time to draw the function to
        an end, having found the shortest string that
        accommodates nothing other than some characters
        in both small and capital forms.
      --]=====]
      
      local temp =
        table.pack( table.unpack( s_list ,
          s_index_start , -1+s_index_start+s_length ) )
      --[=====[
        The table of characters for the current substring in
        question.
      --]=====]
      
      local temp_set = 
        remove_duplicates( temp )
      --[=====[
        Generated a set from the characters within the
        substring. A set rules out duplicates. Will be
        making use of the hashes to check if a 
        character together with its small or capital 
        form is included within the substring in question.
        
        The keys in the table 'temp_set' are without
        duplicates (unique).
      --]=====]
        
      local flag_found = true
      --[=====[
        If the flag above remains true after the 'for' loop
        below, then it will be time to return and yield
        the result.
      --]=====]
      
      for item , _ in pairs( temp_set ) do
        if ( ( mp_small[item] ) and ( temp_set[ mp_small[item] ] ) ) or
           ( ( mp_cap  [item] ) and ( temp_set[ mp_cap  [item] ] ) ) then
          --[=====[
            Making use of the hashes (and drawing
            benefits through the constant-time access)
            to check if a character together with its
            corresponding small or capital form exists
            within the current substring.
          --]=====]
        else
          flag_found = false
          break
          --[=====[
            If the condition above is not satisfied even for
            a single character, then we should break out of the
            for loop since there is no need to lose time
            checking the remaining characters. Setting the
            flag to false will prevent the remaining 'if'
            conditional from terminating the function.
          --]=====]
        end
      end -- for item
      
      if flag_found then
        print( string.format( 
                 "String found in \'%s\' :\n  %s" , 
                 A , table.concat( temp ) ) )
        return #temp
      end
      --[=====[
        The first success to be encountered has the function
        return the length of the string in question.
      --]=====]

    end -- for s_index_start
  end -- for s_length
  
  return -1
  --[=====[
    No possible success meaning there is no substring within
    the original string accommodating both small and capital
    forms of some letters, and nothing else.
  --]=====]
end -- function solution


list_of_strings = {
  "Bb" ,
  "aACCCxA" ,
  "CCzzcZazAxAZ" ,
  "CCzzcybzAxAZ" 
}

for _ , item in ipairs(list_of_strings) do
  print()
  print(
    string.format( "Response to \"%s\" :\n  %s" , 
                   item , solution(item) ) )
end
